File: runtime/java/units/wasi/fd_pwrite.java
// requires: memory/i32_load, memory/i32_store, memory/read_string
// Positional gather write: writes at an absolute file offset without moving the
// channel's own position. Stdio is not seekable (SPIPE).
int wasi_fd_pwrite(int fd, int iovsPtr, int iovsLen, long offset, int nwrittenPtr) {
    Object e = fds.get(fd);
    if (isStdio(e)) {
        return WASI_SPIPE;
    }
    if (!(e instanceof Handle)) {
        return WASI_BADF;
    }
    java.nio.channels.FileChannel ch = ((Handle) e).ch;
    int written = 0;
    try {
        for (int i = 0; i < iovsLen; i++) {
            long base = Integer.toUnsignedLong(iovsPtr) + (long) i * 8;
            int ptr = memory.i32_load(base);
            int len = memory.i32_load(base + 4);
            byte[] chunk = memory.read_string(Integer.toUnsignedLong(ptr), Integer.toUnsignedLong(len));
            java.nio.ByteBuffer bb = java.nio.ByteBuffer.wrap(chunk);
            long pos = offset + Integer.toUnsignedLong(written);
            while (bb.hasRemaining()) {
                pos += ch.write(bb, pos);
            }
            written += len;
        }
    } catch (java.io.IOException ex) {
        return WASI_IO;
    }
    memory.i32_store(Integer.toUnsignedLong(nwrittenPtr), written);
    return WASI_OK;
}

File: runtime/go/units/wasi/fd_pwrite.go
// requires: memory/i32_load, memory/i32_store, memory/read_string
func (w *WASI) wasi_fd_pwrite(fd, iovsPtr, iovsLen uint32, offset uint64, nwrittenPtr uint32) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if w.isStdio(f) {
        return wasiSpipe
    }
    if e := w.checkRight(fd, rightFdWrite); e != wasiOk {
        return e
    }
    written := uint32(0)
    for i := uint32(0); i < iovsLen; i++ {
        ptr := w.memory.i32_load(uint64(iovsPtr) + uint64(i)*8)
        length := w.memory.i32_load(uint64(iovsPtr)+uint64(i)*8 + 4)
        chunk := w.memory.read_string(uint64(ptr), uint64(length))
        // syscall.Pwrite rather than f.WriteAt: WriteAt rejects a fd opened
        // O_APPEND ("invalid use of WriteAt"), and a positional write must ignore append anyway.
        // Portable on darwin+linux.
        n, err := syscall.Pwrite(int(f.Fd()), chunk, int64(offset+uint64(written)))
        written += uint32(n)
        if err != nil {
            return wasiIo
        }
    }
    w.memory.i32_store(uint64(nwrittenPtr), written)
    return wasiOk
}
